### ISSUE TITLE
Slider settings for addons

### ISSUE THREAD

**Issue by DNin01** [ScratchAddons/ScratchAddons#6319]
### Why this would be helpful

Sliders are basically just number inputs but less precise. They're great for numbers and percentages that don't need to be set to an exact value.

### My suggestion

There are multiple different use cases for sliders in addon settings.

#### Percentages and numbers with range restrictions

Sliders work pretty well for percentages, however, some settings deserve fine control, especially in an extension about customization, so there aren't that many places to replace number inputs with sliders.

But there are places where having a number for a setting that seems like a percentage would be very out of place. For example, Microsoft PowerToys has a setting called "Pixel tolerance for edge detection" for Screen Ruler and it goes from 0 to 255 (because this setting is about color, but other than that, it is basically a percentage that affects the sensitivity of edge detection). It would be somewhat awkward to have to adjust it by typing in a new number because the slider makes it easy to make large adjustments within a set range (this is why I hate typing hex codes manually).

Potential use cases:
- #5917
- `progress-bar`
- `middle-click-popup`
- `vol-slider`
- `better-featured-project`
- `zebra-striping`?
- `transparent-orphans`?
- `hi-res-thumbnails`?
- `compact-messages`?

#### Speed settings

A very sensible place for a slider to be is in a multi-choice setting that basically acts as a less precise number, such as an animation speed. These sliders would snap to the nearest setting (represented by a tick mark) when released. I showed this in #6317.

**Comment by Secret-chest** [comment]
It would make sense only if there was always a typeable number input next to the slider.

**Comment by BroJac5246** [comment]
> It would make sense only if there was always a typeable number input next to the slider.

THIS.

I hate sliders, but this would be great!

**Comment by WorldLanguages** [comment]
Yeah, this would be an optional addition for number settings, and users will still be able to see and type in numbers.

For some "select" settings which can be ordered, the slider would effectively be the only way to change the setting, such as speed:slow/normal/fast.

**Comment by Secret-chest** [comment]
> For some "select" settings which can be ordered, the slider would effectively be the only way to change the setting, such as speed:slow/normal/fast.

What about making the speed a percentage, and having presets for it in the reset button?

**Comment by DNin01** [comment]
> > For some "select" settings which can be ordered, the slider would effectively be the only way to change the setting, such as speed:slow/normal/fast.
> 
> What about making the speed a percentage, and having presets for it in the reset button?

Probably a bit too precise. For example, on #4803, the initial idea was to add a setting for time in milliseconds, but it turned out later that three options for the refresh speed was more user-friendly.

This extension isn't meant to be the most customizable thing ever, but it has enough options for most users.

**Comment by Secret-chest** [comment]
> Probably a bit too precise.

It's a win-win if we make it easier to use _and_ more precise.

**Comment by DNin01** [comment]
> > Probably a bit too precise.
> 
> It's a win-win if we make it easier to use _and_ more precise.

Well, there is software with UI that augments slider controls with number inputs so you can type in values or slide to make coarse adjustments.

Even then, not everything has to be able to be adjusted fine amounts, and having a number input to the side doesn't necessarily make the setting easier to use in every case. We want to make a user-friendly extension and not very many users will care about transition durations down to the millisecond. When tiny changes aren't noticeable, it doesn't matter if the slider isn't that precise. For example, how would you like customizing exactly how long it takes to open a context menu on a touch screen by pressing and holding when (as an example) you don't know what a millisecond is, when you could've just had a simple choice between "short", "medium", and "long", without having to figure out any numbers? (Well, for that, it might depend on the context, I guess.)

I'm not saying there aren't cases where a slider and an input together wouldn't make sense, it's just important not to overuse settings with this level of control.

**Comment by Secret-chest** [comment]
> For example, how would you like customizing exactly how long it takes to open a context menu on a touch screen by pressing and holding when (as an example) you don't know what a millisecond is, when you could've just had a simple choice between "short", "medium", and "long", without having to figure out any numbers? (Well, for that, it might depend on the context, I guess.)

FIY Samsung uses a simulator. You just hold down how much you like and that's your delay.

* * *

On a desktop OS, the double-click delay is set with a slider, and it's easy to understand.

It's not like supporting in-between values makes our lives harder.

**Comment by DNin01** [comment]
> FIY Samsung uses a simulator. You just hold down how much you like and that's your delay.

Interesting but could take a moment to understand.

> On a desktop OS, the double-click delay is set with a slider, and it's easy to understand.

Sure, but we'd need a test, which, as an example, iOS does with Haptic Touch (named after the haptic feedback you get when pressing and holding to open a context menu), you select between three speed options and an icon is shown below the setting that you can long press to expand, therefore you get to test each setting.

It would also be cool if this was a sort of "global setting" that, when set, applied to all addons. A setting like that would probably have to be in a modal opened by clicking a button on the addon's settings, or we could tell the user where to go to change it.

**Comment by BroJac5246** [comment]
> > FIY Samsung uses a simulator. You just hold down how much you like and that's your delay.
> 
> Interesting but could take a moment to understand.
> 
> > On a desktop OS, the double-click delay is set with a slider, and it's easy to understand.
> 
> Sure, but we'd need a test, which, as an example, iOS does with Haptic Touch (named after the haptic feedback you get when pressing and holding to open a context menu), you select between three speed options and an icon is shown below the setting that you can long press to expand, therefore you get to test each setting.
> 
> It would also be cool if this was a sort of "global setting" that, when set, applied to all addons. A setting like that would probably have to be in a modal opened by clicking a button on the addon's settings, or we could tell the user where to go to change it.

You guys are overthinking it. As a user, it's a lot of effort to manually choose a long-press delay. I'd rather just stick with the system default that I'm used to.

**Comment by Secret-chest** [comment]
I always double-click fast. This is more about the new addon to double-click blocks; I want to use a smaller delay.

**Comment by BroJac5246** [comment]
> I always double-click fast. This is more about the new addon to double-click blocks; I want to use a smaller delay.

A long period for which it counts as a double-click still accommodates people who double-click quickly.

**Comment by Secret-chest** [comment]
> A long period for which it counts as a double-click still accommodates people who double-click quickly.

What if you accidentally click often?

**Comment by BroJac5246** [comment]
> > A long period for which it counts as a double-click still accommodates people who double-click quickly.
> 
> What if you accidentally click often?

But why would you do that? Maybe if you have a disability?

**Comment by Secret-chest** [comment]
> But why would you do that? Maybe if you have a disability?

Touchpad users... Also, I use a mouse and randomly click to ensure the workspace is focused.
